tests/cpp/test_move_pad.cpp

```diff
@@ -0,0 +1,437 @@
+
+  testValidate(fec.fusion(), out_tensors, aten_inputs, __LINE__, __FILE__);
+}
+
+TEST_F(MovePadTest, PadReplayOnMultipleUsesCase0) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* tv0 = makeContigConcreteTensor({4, 10});
+  TensorView* tv1 = makeContigConcreteTensor({1, 10});
+  // pad on tv5 will propagation back to tv0, since all its uses are traversed
+  //
+  //   tv0 --> tv2 --> tv3 --> tv5
+  //               \        /
+  //                -> tv4 -
+  TensorView* tv2 = relu(tv0);
+  TensorView* tv3 = neg(tv2);
+  TensorView* tv4 = sin(tv2);
+  TensorView* tv5 = add(tv3, tv4);
+  TensorView* tv6 = cat({tv5, tv1}, /*dim=*/0);
+
+  fusion->addInput(tv0);
+  fusion->addInput(tv1);
+  fusion->addOutput(tv6);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor t0 = at::randn({4, 10}, options);
+  at::Tensor t1 = at::randn({1, 10}, options);
+  std::vector<c10::IValue> aten_inputs = {t0, t1};
+
+  FusionExecutorCache fec(std::move(fusion));
+  auto out_tensors = fec.runFusionWithInputs(aten_inputs);
+
+  FusionKernelRuntime* runtime = fec.getMostRecentKernelRuntime();
+  EXPECT_EQ(runtime->fusionSegments()->groups().size(), 1);
+
+  testValidate(fec.fusion(), out_tensors, aten_inputs, __LINE__, __FILE__);
+}
+
+TEST_F(MovePadTest, PadReplayOnMultipleUsesCase1) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* tv0 = makeContigConcreteTensor({4, 10});
+  TensorView* tv1 = makeContigConcreteTensor({4, 10});
+  // pad on tv5 will NOT propagation back to tv0, since there's one path
+  //   tv0 --> tv2 --> tv7
+  // which is not covered by reverse traversing through producers of tv5.
+  // So the pad propagation would stop at tv2
+  TensorView* tv2 = relu(tv0);
+  TensorView* tv3 = neg(tv2);
+  TensorView* tv4 = sin(tv2);
+  TensorView* tv5 = add(tv3, tv4);
+  TensorView* tv6 = cat({tv5, tv1}, /*dim=*/0);
+  TensorView* tv7 = sin(tv2);
+
+  fusion->addInput(tv0);
+  fusion->addInput(tv1);
+  fusion->addOutput(tv6);
+  fusion->addOutput(tv7);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor t0 = at::randn({4, 10}, options);
+  at::Tensor t1 = at::randn({4, 10}, options);
+  std::vector<c10::IValue> aten_inputs = {t0, t1};
+
+  FusionExecutorCache fec(std::move(fusion));
+  auto out_tensors = fec.runFusionWithInputs(aten_inputs);
+
+  testValidate(fec.fusion(), out_tensors, aten_inputs, __LINE__, __FILE__);
+}
+
+TEST_F(MovePadTest, CascadePadCase0) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  // all pad operations should be merged together
+  TensorView* tv0 = makeContigConcreteTensor({4, 10});
+  TensorView* tv1 =
+      pad(tv0,
+          {IrBuilder::create<Val>(0L),
+           IrBuilder::create<Val>(2L),
+           IrBuilder::create<Val>(1L),
+           IrBuilder::create<Val>(1L)});
```

Comment:
(Can come as a separate PR)

It feels worth adding to csrc/ops/alias.h a helper that takes integer pad widths.